web/src/composebox_typeahead.js

```diff
@@ -151,38 +152,70 @@ export function should_enter_send(e) {
     return this_enter_sends;
 }
 
+function handle_bulleting_or_numbering($textarea, e) {
+    // handles automatic insertion or removal of bulleting or numbering
+    const before_text = split_at_cursor($textarea.val(), $textarea)[0];
+    const previous_line = list_util.get_last_line(before_text);
+    let to_append = "";
+    // if previous line was bulleted automatically add a bullet to the new line
+    if (list_util.is_bulleted(previous_line)) {
+        // if previous line had only bullet, remove it and stay on the same line
+        if (list_util.strip_bullet(previous_line) === "") {
+            // the below line replaces the last 2 characters in the textarea before
+            // the cursor - the bullet syntax - with an empty string
+            $textarea.range($textarea.caret() - 2, $textarea.caret()).range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        // use same bullet syntax as the previous line
+        to_append = previous_line.slice(0, 2);
+    } else if (list_util.is_numbered(previous_line)) {
+        // if previous line was numbered continue numbering with the new line
+        const previous_number_string = previous_line.slice(0, previous_line.indexOf("."));
+        // if previous line had only numbering, remove it and stay on the same line
+        if (list_util.strip_numbering(previous_line) === "") {
+            // the below line replaces the last few characters in the textarea before
+            // the cursor - the numbering syntax - with an empty string
+            $textarea
+                .range($textarea.caret() - previous_number_string.length - 2, $textarea.caret())
+                .range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        const previous_number = Number.parseInt(previous_number_string, 10);
+        to_append = previous_number + 1 + ". ";
+    }
+    // if previous line was neither numbered nor bulleted, only add
+    // a new line to emulate default behaviour (to_append is blank)
+    // else we add the bulleting / numbering syntax to the new line
+    $textarea.caret("\n" + to_append);
```

Comment:
I think this particular place lacking `text-field-edit` is highly problematic -- it means using `enter` breaks undo every time, not just when using the new formatting buttons.

Reply:
Changed to `text-field-edit`'s `insert`